flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/Table.scala

```diff
@@ -1,271 +1,291 @@
+   * }}}
+   */
+  def groupBy(fields: Expression*): Table = {
+    val analyzer = new GroupByAnalyzer(operation.outputFields)
+    val analyzedFields = fields.map(analyzer.analyze)
+
+    val illegalKeys = analyzedFields filter {
+      case fe: ResolvedFieldReference => false // OK
+      case e => true
+    }
+
+    if (illegalKeys.nonEmpty) {
+      throw new ExpressionException("Illegal key expressions: " + illegalKeys.mkString(", "))
+    }
+
+    this.copy(operation = GroupBy(operation, analyzedFields))
+  }
+
+  /**
+   * Groups the elements on some grouping keys. Use this before a selection with aggregations
+   * to perform the aggregation on a per-group basis. Similar to a SQL GROUP BY statement.
+   *
+   * Example:
+   *
+   * {{{
+   *   in.groupBy("key").select("key, value.avg")
+   * }}}
+   */
+  def groupBy(fields: String): Table = {
+    val fieldsExpr = ExpressionParser.parseExpressionList(fields)
+    groupBy(fieldsExpr: _*)
+  }
+
+  /**
+   * Joins two [[Table]]s. Similar to an SQL join. The fields of the two joined
+   * operations must not overlap, use [[as]] to rename fields if necessary. You can use
+   * where and select clauses after a join to further specify the behaviour of the join.
+   *
+   * Example:
+   *
+   * {{{
+   *   left.join(right).where('a === 'b && 'c > 3).select('a, 'b, 'd)
+   * }}}
+   */
+  def join(right: Table): Table = {
+    val leftInputNames = operation.outputFields.map(_._1).toSet
+    val rightInputNames = right.operation.outputFields.map(_._1).toSet
+    if (leftInputNames.intersect(rightInputNames).nonEmpty) {
+      throw new ExpressionException(
+        "Overlapping fields names on join input, result would be ambiguous: " +
+          operation.outputFields.mkString(", ") +
+          " and " +
+          right.operation.outputFields.mkString(", ") )
+    }
+    this.copy(operation = Join(operation, right.operation))
+  }
+
+  /**
+   * Union two[[Table]]s. Similar to an SQL UNION ALL. The fields of the two union operations
+   * must fully overlap.
+   *
+   * Example:
+   *
+   * {{{
+   *   left.unionAll(right)
+   * }}}
+   */
+  def unionAll(right: Table): Table = {
+    val leftInputFields = operation.outputFields
+    val rightInputFields = right.operation.outputFields
+    if (!leftInputFields.equals(rightInputFields)) {
+      throw new ExpressionException(
+        "The fields names of join inputs should be fully overlapped, left inputs fields:" +
+          operation.outputFields.mkString(", ") +
+          " and right inputs fields" +
+          right.operation.outputFields.mkString(", ")
+      )
+    }
+    this.copy(operation = UnionAll(operation, right.operation))
+  }
+
+  /**
+   * Get the process of the sql parsing, print AST and physical execution plan.The AST
+   * show the structure of the supplied statement. The execution plan shows how the table 
+   * referenced by the statement will be scanned.
+   */
+  def explain(extended: Boolean): String = {

```

Comment:
It looks like you add only this method.
Can you check why the whole file is marked as changed?